Sonifcation/Function.cs

```diff
@@ -107,34 +139,66 @@ public partial class Function : Node {
 	/// </summary>
 	private void FillFunctionTable() {
 
-		// Iterate over the Domain and fill in the function's range
-		for(double t = StartTime; t <= EndTime; t += NoteDuration) {
+		// Iterate over the Domain and fill in the Function range, taking special care to tag Function as constant if needed.
+		IsConstant = true;
+		double lastValue = FunctionAST.EvaluateAtT(StartTime);
+		FunctionTable.Add(lastValue);
+
+		for(double t = StartTime + NoteDuration; t <= EndTime; t += NoteDuration) {
+
+			// Evalauate the funtion at the approriate time.
 			var value = FunctionAST.EvaluateAtT(t);
+			
+			// Manage constant/nonconstant function case.
+			if(IsConstant) {
+				if(value != lastValue) IsConstant = false;
+				else lastValue = value;
+			}
+
+			// Manage out of bounds case.
+			if(Math.Abs(value) == double.PositiveInfinity) value = 0;
+
+			// Add the value to the functions table.
 			FunctionTable.Add(value);
 		}
+
 	}
 	
 	/// <summary>
 	/// Uses Function <c>FunctionTable</c> to calculate the appropriate values for the functions audio sequence.
 	/// </summary>
 	private void FillNoteSequence() {
 
-		// Functions have 88 notes to choose from.
-		int noteNumStart = 1;
-		int noteNumEnd = 88;
+		// Functions have a set amount of notes to choose from defined impliclty in 'Frequencies.cs'.
+		int noteNumStart = Frequencies.startingNoteNumber;
+		int noteNumEnd = Frequencies.endingNoteNumber;
 		int noteRange = noteNumEnd - noteNumStart;
 		
 		// Find the full range of values that the function can take.
 		double minVal = FunctionTable.Min();
 		double maxVal = FunctionTable.Max();
 		double functionRange = maxVal - minVal;
-
-		// Map the Functions values to note numbers 
+		
+		// Map the Functions values to note numbers
 		foreach(double value in FunctionTable) {
-			double normalizedValue = (value - minVal) / functionRange;
-			int note = noteNumStart + (int) (normalizedValue * noteRange);
-			noteSequence.Add(note);
+			int noteNumber;
+			if(IsConstant) noteNumber = (int) Math.Abs(FunctionTable[0]) % noteRange;
+			else {
+				double normalizedValue = (value - minVal) / functionRange;
```

Comment:
Normalizing the function range doesn't seem like the way to go; this way `x + 3` and `x + 5` would sound exactly the same even though the latter is higher than the former. Unless I'm mistaken on how this works?

Reply:
I would expect it to map the output of the function to the note range in the same way it maps constant functions now. Then you can just stop handling constant functions explicitly at all

Reply:
They will sound the same, that is the primary limitation of doing it in this way. If I used a modified version of how I handled the constant functions, it would look something like `noteNumber = (int) (Math.Abs(value) + Frequencies.tuningKeyNum) % noteRange;` (provided I make `tuningKeyNum` public). Doing this would create a Note Space for the function that looks like and sounds like this:
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/9563f6dc-69cb-4b96-8159-751633bd8e46)

as opposed to looking and sounding closer to this when using a linear mapping:
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/ca830b9d-68ee-42c8-b3e9-4d56e756f12a)

Your suggestion is what I originally had in mind. I stopped trying to work it that way because the notes were just too discrete and I moved to try and get a functional audio interface using the linear mapping instead with the idea being that every function sounds differently with respect to its own space. A way to mitigate the limitation might be to add a small but significant bias to similar functions with identical ranges (as this limitation only shows up in this way) and by doing so make sure their output isn't identical. This is a band aid solution though and I am willing to go back and reevaluate how I could make the original idea work and find a different way to calculate the proper audio output.

Reply:
Why not just increase the vertical resolution with the constant method? You can use `%` with floats to directly map any sample into a reasonable range without clamping it to integer values, so you'd still get a smooth mapping as with linear instead of discrete steps.

Reply:
That was part of the original idea I ended up scrapping. I think the way you described it is the same way I had it implemented while fleshing things out in the beginning. Doing that would result in modifying the `FillNoteSequence()` method to look something like this:

```
double startFreq = Frequencies.Notes[Frequencies.startingNoteNumber];
double endFreq = Frequencies.Notes[Frequencies.endingNoteNumber];
noteNumber = (Math.Abs(value) + (int) Frequencies.tuningFreq + 1) % (endFreq) + startFreq;
```

This takes the same kind of logic I used to play constant functions and applies it to a frequency space stripped of notes. Here `noteNumber` is basically just the frequency to generate audio for and you would adjust the 'GenerateNoteAudio()' method accordingly to accommodate for this change in meaning.  `noteNumber` is also biased by translating the Functions f(t) values by adding to them `Frequencies.tuningFreq` so that calculated frequencies can guarantee that a meaningfully audible noise will be played (if I didn't do this, frequencies calculated in this manner would be low and borderline inaudible due to how tiny f(t) might be  function. This scheme produces a frequency space like so, of course linear like the direct linear mapping technique: 
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/dc89dbe2-22d3-4dbb-af9a-74503ad79d27)

The problem with this approach lies in that the range of sound played this way is so small that the difference in tones as we transition from beginning to end is just barely noticeable. The example graph above shows only a difference of 10 Hz between the initial and final notes played and it doesn't get much better for other functions unless you have a much larger range in the time domain. If you want to try it out and see how it sounds, I've updated the branch with this method and you'd just have to load it and press the play button on the toolbar. Further explained in the most recent commit is how to switch between methods of playing the audio if you want to compare all three we've discussed (Direct Linear Mapping, Mapping Directly to the Notes and Rounding, Mapping Directly to Frequency Space w/ No Rounding).

Reply:
I was thinking more like this (pseudocode):
`noteNumber = (Function.At(t) + C5) % 88.0;`
This way 0 would always be C5, every 1 y value goes up 1 note or down 1 note, looping back to the other end of the scale at the ends. Rather than grabbing the min and max notes and assigning those as the frequency space of the function.

Reply:
Funnily enough, that's pretty much the same thing I wrote in case 2 of the Function class `FillNoteSequence()` method but a more basic version:
```
case 2:
	double note = (Math.Abs(value) + Frequencies.tuningKeyNum) % noteRange + Frequencies.startingNoteNumber;
	// Round up if necessary.
	if(Math.Ceiling(note) - note <= 0.5) noteNumber = Math.Ceiling(note);
	else noteNumber = Math.Floor(note);
	break;
```

The difference between this version and the pseudocode is that this takes into account the possibility of starting form different notes other than 1. Using this with the piano notes scheme (88 discrete notes mapped to frequencies), we have to round up or truncate that results. Kind of talked about that earlier in this thread if you want to read more about it again and why I thought we might need to find a better way to sonify.